src/EditorFeatures/Core/GoToDefinition/AbstractGoToDefinitionService.cs

```diff
@@ -40,10 +40,31 @@ internal abstract class AbstractGoToDefinitionService : AbstractFindDefinitionSe
         async Task<IEnumerable<INavigableItem>?> IGoToDefinitionService.FindDefinitionsAsync(Document document, int position, CancellationToken cancellationToken)
             => await FindDefinitionsAsync(document, position, cancellationToken).ConfigureAwait(false);
 
+        private async Task<bool> TryNavigateToSpanAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            // Can only navigate or present items on UI thread.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var solution = document.Project.Solution;
+            var workspace = solution.Workspace;
+            var service = workspace.Services.GetRequiredService<IDocumentNavigationService>();
+
+            var options = solution.Options.WithChangedOption(NavigationOptions.PreferProvisionalTab, true);
+            options = options.WithChangedOption(NavigationOptions.ActivateTab, true);
+
+            return service.TryNavigateToPosition(workspace, document.Id, position, virtualSpace: 0, options, cancellationToken);
+        }
+
         public bool TryGoToDefinition(Document document, int position, CancellationToken cancellationToken)
         {
-            // Try to compute the referenced symbol and attempt to go to definition for the symbol.
             var symbolService = document.GetRequiredLanguageService<IGoToDefinitionSymbolService>();
+            var targetPositionOfControlFlow = symbolService.GetTargetIfControlFlowAsync(document, position, cancellationToken).WaitAndGetResult(cancellationToken);
+            if (targetPositionOfControlFlow is not null)
+            {
+                return _threadingContext.JoinableTaskFactory.Run(() => TryNavigateToSpanAsync(document, targetPositionOfControlFlow.Value, cancellationToken));
```

Comment:
@sharwell for jtf review.

Reply:
The preceding call to `WaitAndGetResult` requires the main thread, so we know at this point we are still on the main thread. Easiest fix without rewriting this type to correctly use async/await is to make `TryNavigateToSpanAsync`&rarr;`TryNavigateToSpan` synchronous and just call it directly here.

Reply:
Done. Thanks for the advice